service/src/utils/onchain-data.ts

```diff
@@ -45,13 +43,30 @@ export async function getOnchainCursedSarcophagi(web3Interface: Web3Interface):
     const sarcoIds = await web3Interface.viewStateFacet.getArchaeologistSarcophagi(web3Interface.ethWallet.address);
     sarcoIds.map(async sarcoId => {
         const sarco = await web3Interface.viewStateFacet.getSarcophagus(sarcoId);
+        const archStorage = await web3Interface.viewStateFacet.getSarcophagusArchaeologist(sarcoId, web3Interface.ethWallet.address);
 
-        if (sarco.state === SarcophagusState.Exists) {
-            // only add sarco that have state === EXISTS
+        if (sarco.state === SarcophagusState.Exists && !archStorage.unencryptedShard) {
+            const nowTimestamp = (new Date()).getUTCSeconds();
+            // TODO: rename resurrectionWindow to gracePeriod when contract updates merged
+            const tooLateToUnrap = sarco.resurrectionTime.toNumber() + sarco.resurrectionWindow.toNumber() < nowTimestamp;
+
+            if (tooLateToUnrap) {
+                archLogger.warn(`You failed to unwrap a Sarcophagus ${sarcoId}\n`);
+                // TODO: archaeologist might want to call clean here
+
+                return;
+            }
+
+            // only add unwrappable sarco that have state === EXISTS and haven't already unwrapped (uploaded their unencrypted shard)
+            const resurrectionTime = new Date(sarco.resurrectionTime.toNumber() * 1000);
             archSarco.push({
                 id: sarcoId,
-                resurrectionTime: new Date(sarco.resurrectionTime.toNumber() * 1000),
+                resurrectionTime,
             });
+
+            // Schedule an uwrap job for each sarco this arch is cursed on. `scheduleUnwrap` will cancel existing 
+            // schedules, so no duplicate jobs will be created.
+            scheduleUnwrap(web3Interface, sarcoId, resurrectionTime);
```

Comment:
Will this work if the job is scheduled in the past?

Reply:
Tests showed that scheduling in the past (or even on the exact second) won't work, so in that case I'm making an immediate call to `unwrapSarcophagus` without scheduling anything.